### PR TITLE
R20.2 312

### PR DESCRIPTION
**Pull request by dougkenik**
Fix for #312, #317, #266, and #304 

For #312 - I decided the best way to solve this is to make any modifier meshes which intersect the normal mesh a child of the normal mesh when the project is saved.

### CODE REVIEW

**Review by jake-kirkland-teton — APPROVED** [review]
Changes look good. The bugs appear fixed, as long as the save dialog appears correctly for Brady.

File: SmartSlicePlugin/SmartSliceSavePrompt.qml [hunk 1]
```diff
@@ -15,15 +15,18 @@ UM.Dialog {
...
 
     minimumWidth: width;
     maximumWidth: width;
 
     minimumHeight: height;
     maximumHeight: height;
 
+    x: 0.5 *(CuraApplication.appWidth() - width)
+    y: 0.5 *(CuraApplication.appHeight() - height)
+
```
Comment (colman-hartinger) [inline SmartSlicePlugin/SmartSliceSavePrompt.qml:29]: Add space after *

**Review by colman-hartinger — CHANGES_REQUESTED** [review]
Code looks good and bug seems fixed. Just one quick styling change and should be good to go

**Review by colman-hartinger — APPROVED** [review]

File: SmartSlicePlugin/SmartSliceProperty.py [hunk 2]
```diff
@@ -285,7 +286,7 @@ def __init__(self, node: CuraSceneNode = None):
 
     def value(self):
         if self._node:
-            return self._node.getScale(), self._node.getOrientation(), self._node.getPosition(), self._node.getDecorator(ZOffsetDecorator), self.findIntersectingNodes()
+            return self._node.getScale(), self._node.getOrientation(), self._node.getPosition(), self._node.getDecorator(ZOffsetDecorator), intersectingNodes(self._node)
```
Comment (bradymadams) [inline SmartSlicePlugin/SmartSliceProperty.py:289]: Break up long line

**Review by bradymadams — CHANGES_REQUESTED** [review]
The save dialog looks good on my machine now.

I have some questions about checking the intersecting nodes and setting parents.

1. Does this have the potential to set a parent, or set a node's parent to a different node when we wouldn't want to? Mainly I'm thinking about someone using Cura for a non Smart Slice workflow - is it possible we'd manipulate the scene in a way they wouldn't want us to when they save the 3MF?
2. Will this be a non-issue for any 3MF saved with the fix from https://github.com/tetonsim/is-cura-ui/pull/334

**Comment by dougkenik** [comment]
> The save dialog looks good on my machine now.
> 
> I have some questions about checking the intersecting nodes and setting parents.
> 
> 1. Does this have the potential to set a parent, or set a node's parent to a different node when we wouldn't want to? Mainly I'm thinking about someone using Cura for a non Smart Slice workflow - is it possible we'd manipulate the scene in a way they wouldn't want us to when they save the 3MF?
> 2. Will this be a non-issue for any 3MF saved with the fix from #334

1. Yes, if the Plugin is installed, and a user saves, then we could set or change the parent. I went back and forth on this, and it seems like the best option to solve this bug (I have no idea how to do it otherwise). I also have no clue how a SceneNode can get set as a parent or child in any other Cura workflow (besides, maybe, grouping)
2. No - this is a very different problem, and happens on project open for a modifier mesh which does not have a parent. The modifier mesh is always dropped to the build plate if it does not have a parent. For optimizations, we set a parent for the mod mesh, so it gets read in at the correct location and will not get dropped to the build plate (which is why I chose to go this route)

**Comment by bradymadams** [comment]
Is the issue with manually added modifier meshes only? If so, maybe we can check for modifier meshes when a validation starts and if they're parent is the root reset it to the one printable node. Would that work?

**Comment by dougkenik** [comment]
> Is the issue with manually added modifier meshes only? If so, maybe we can check for modifier meshes when a validation starts and if their parent is the root reset it to the one printable node. Would that work?

This is the case. We could do that, but then there is an issue that the modifier meshes would not move independently like when they were brought in. I chose to do this on project save because then the UI appears to be consistent if the user is iterating on moving modifier meshes. Once the mod mesh gets a parent, the mod mesh is moved automatically when the parent is moved, but the mod mesh can be moved without the parent being moved. The other issue is that if a user sets up a validation, does not run, and saves, the modifier mesh could be dropped to the build plate when they open again

**Comment by bradymadams** [comment]
Okay, well, I really don't like that this changes the scene of non Smart Slice workflows. I think we need to figure out a different approach.

**Comment by dougkenik** [comment]
> Okay, well, I really don't like that this changes the scene of non Smart Slice workflows. I think we need to figure out a different approach.

Ok... Any way we automatically solve this problem will result in the scene changing. Even if we find a way to not set the parent of the modifier mesh, on project open we would need to adjust the Z-location of the modifier mesh (which I do not think is feasible from what I've seen when we open 3MFs). We could set the parent of the modifier mesh when the user enters the stage or has any interaction in the stage - that would indicate the user has "intent" to use Smart Slice, and we would need to remember the position of the modifier mesh.

The other way would be to add the modifier mesh to our Smart Slice metadata in the Cura project and re-insert it when they enter the stage... it's not elegant though, and we would need to remove it and insert the Cura one if they leave the stage

**Comment by bradymadams** [comment]
> We could set the parent of the modifier mesh when the user enters the stage or has any interaction in the stage - that would indicate the user has "intent" to use Smart Slice, and we would need to remember the position of the modifier mesh.

I think this is definitely better. And it would remove all ambiguity if we prompted the user before doing changing the node's parent.

**Comment by dougkenik** [comment]
At this point, we might as well just prompt the user on project save. If they decline, it's on them

**Comment by bradymadams** [comment]
> At this point, we might as well just prompt the user on project save. If they decline, it's on them

I guess I'm fine with either since they both require user confirmation before changing the scene.

**Comment by dougkenik** [comment]
> > At this point, we might as well just prompt the user on project save. If they decline, it's on them
> 
> I guess I'm fine with either since they both require user confirmation before changing the scene.

Ok, I moved all of the logic to `onStageSelected` and will issue a dismissible message when we change the parent of a mod mesh.

**Review by jake-kirkland-teton — APPROVED** [review]
Changes look good. The bugs appear fixed, as long as the save dialog appears correctly for Brady.

**Review by bradymadams — APPROVED** [review]
Will merge pending approval from @colman-hartinger

**Review by colman-hartinger — APPROVED** [review]

**Pull request merged**